ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -1088,6 +1127,33 @@ void addChunkId(BlobId chunkBlobId, int chunkIndex) {
       indexToChunkIds.put(chunkIndex, chunkBlobId);
     }
 
+    /**
+     * Call {@link NotificationSystem#onBlobCreated(String, BlobProperties, byte[], NotificationBlobType)} for this
+     * chunk, unless it is the first chunk, in which case it might be an entire simple blob. In that case, save
+     * the {@link BlobProperties} from the first chunk.
+     * @param chunk the {@link PutChunk} created.
+     */
+    void maybeNotifyForChunkCreation(PutChunk chunk) {
+      if (chunk.chunkIndex == 0) {
+        firstChunkIdAndProperties = new Pair<>(chunk.chunkBlobId, chunk.chunkBlobProperties);
+      } else {
+        notificationSystem.onBlobCreated(chunk.chunkBlobId.getID(), chunk.chunkBlobProperties, userMetadata,
+            NotificationBlobType.DataChunk);
+      }
+    }
+
+    /**
+     * Notify for the creation of the first chunk. To be called after the overall operation is completed if the overall
+     * blob is composite. If no first chunk was put successfully, this will do nothing.
+     */
+    void notifyForFirstChunkCreation() {
+      if (firstChunkIdAndProperties != null) {
```

Comment:
Maybe omit this check since this method will now only be called if it is known for sure that there is a first chunk that needs notification (which is also why it is no longer prefixed with "maybe").